File: #1038.binary-search-tree-to-greater-sum-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bstToGst(TreeNode* root) {

        if (root) {
           self:bst(root);
        }

        return root;
    }
    
private:
    int order = 0;

    void bst(TreeNode* node) {
    
        if (node->right) {
            bst(node->right);
        }
        
        node->val += order;
        order = node->val;
        
        if (node->left) {
            bst(node->left);
        }
    }
};

File: #1038.binary-search-tree-to-greater-sum-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var bstToGst = function(root) {
    let sum = 0; 
    var sums = (node) => {
        if ( node === null) {
            return null;
        }

        sums(node.right);
        sum = sum + node.val;
        node.val = sum;
        sums(node.left)
    };
    sums(root);
    return root;
};
